Add new Config function to register templates

eileenmcnaughton opened the pull request:
I had been upgrading some of the extensions civix to use the addTemplateDir method (this is already merged to civix and replaces the more cludgey way that fails post Smarty2).

However, it struck me that we are still calling a
class that extends the Smarty class - which means we have fewer options than if we call a separate class, with no Smarty dependency.

The specific challenge currently is that
loading version 3 from an extension 'works' but I have to directly hack the require_once in the file because the CRM_Core_Smarty file is loaded by the first class that attempts to run it's stock civix config. In theory if the classes didn't do this then the extension's auto-loader would kick in and we could do a class_exists check first. I'm sure it's more difficult than that but
I think not involving the Smarty class directly
in our config functions we have more options
to evolve it.

I made the signature an array since that felt more flexible although only type=Smarty is a thing and
the version is ignored at this stage

Note this would also require a civix update & the goal is to start phasing this in in preparation for a future option to use Smarty3

Related https://github.com/civicrm/civicrm-core/pull/25248

Overview
----------------------------------------
_A brief description of the pull request. Keep technical jargon to a minimum. Hyperlink relevant discussions._

Before
----------------------------------------
_What is the old user-interface or technical-contract (as appropriate)?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

After
----------------------------------------
_What changed? What is new old user-interface or technical-contract?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

Technical Details
----------------------------------------
_If the PR involves technical details/changes/considerations which would not be manifest to a casual developer skimming the above sections, please describe the details here._

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

eileenmcnaughton committed:
Update Smarty addTemplateDir function signature to future smarty

eileenmcnaughton committed:
Add new Config function to register templates

I had been upgrading some of the extensions civix to use the
addTemplateDir method (this is already merged to civix
and replaces the more cludgey way that fails post Smarty2).

However, it struck me that we are still calling a
class that extends the Smarty class - which means we have
fewer options than if we call a separate class, with no
Smarty dependency.

The specific challenge currently is that
loading version 3 from an extension 'works' but I have to
directly hack the require_once in the file because the
CRM_Core_Smarty file is loaded by the first class that
attempts to run it's stock civix config. In theory if
the classes didn't do this then the extension's auto-loader
would kick in and we could do a class_exists check
first. I'm sure it's more difficult than that but
I think not involving the Smarty class directly
in our config functions we have more options
to evolve it.

I made the signature an array since that felt more flexible
although only type=Smarty is a thing and
the version is ignored at this stage

Note this would also require a civix update & the goal is to
start phasing this in in preparation for a future option to
use Smarty3

Related https://github.com/civicrm/civicrm-core/pull/25248

eileenmcnaughton commented:
test this please

totten commented:
@eileenmcnaughton Yeah, I agree that overriding methods on `Smarty` is a messy situation.

I like the idea of separating out the interface for this, but there is an issue around compatibility - e.g. the new entry-point wouldn't be available until (say) 5.58. Since `civix` generally targets a slightly older runtime, that means `civix` would either need to hold back for a while -- or use some kind of polyfill/adapter.

What if we shift to a mixin? For example:

https://github.com/civicrm/civicrm-core/compare/master...totten:civicrm-core:master-smarty-mixin?expand=1

In that approach:

* `smarty-v2@1.0.0` would be the initial version. It would be written in a way that's compatible with older versions of Civi. We would [tell civix about it](https://github.com/totten/civix/blob/master/mixin-backports.php). When generating code (with older release targets), civix would add this.
* Going forward, if we need to change `smarty-v2` to use a different functions, then we can update `civicrm-core.git:mixin/smarty-v2@1/mixin.php`. (*Extensions don't need any more updates.*)
* For your extension that experimentally uses Smarty v3, you could have an even newer version. `ext/smarty3/mixin/smarty-v2@1.99.0.mixin.php` would take precedence. Then you could do some v3-specific tricks to use a v3-runtime with the v2 data-folders.

(*Aside: Not sure... maybe the name `smarty@2.0.0` would be more sensible than `smarty-v2@1.0.0`. The mechanics would mostly work the same, but the first could be misinterpreted as marking the Smarty version as 2.0.0.*)

eileenmcnaughton committed:
Update Smarty mixin to meet initial goal

This gets us away from firing the Smarty object directly - which gives us
greater flexibilty with loading the Smarty object. It also uses the preferred function
and addresses the deduping that seemed to be the issue

eileenmcnaughton commented:
@totten I've updated this to build on the work you did to get to the original goal of the piece of work ie

1) stop instantiating the `Smarty` object directly - since as soon as we do that it jumps in on the `require_once` - which  I was looking to have more control over. In my testing I was able to get `Smarty3` running largely by changing the `require_once` (and fixing some of our other bits & pieces). It could be done in my test `Smarty3` extension with the additon of a `IF (!class_exists`I don't know if this will be our final approach but the intention was to separate the loading from the need to register a template to give more options, especially in the experimental phase.

2) Fix the signature of `addTemplate` to be forward compatible with future smarty versions - that was in a separate PR https://github.com/civicrm/civicrm-core/pull/25248 but I have included here as well - although that PR could be merged as the feedback had been addressed in which case I can rebase out

**Note** I was unable to get `civix upgrade` to run with the smarty mixing after doing a git pull on `civix` and on `civicrm-buildkit` - even on Civi Grant - so I have not managed to do that side of it.

Enable mixin smarty-v2@1.0.0

In Mixlib.php line 184:
                                                                                                               
  Failed to locate smarty-v2@1/mixin.php (mixlibDir=/srv/civicrm-buildkit/bin/civix/extern/mixin, mixlibUrl=)

totten committed:
(NFC) mixer - Fix PHP 8.1 warning

totten commented:
> Note I was unable to get `civix upgrade` to run with the smarty mixing after doing a git pull on `civix` and on `civicrm-buildkit` - even on Civi Grant - so I have not managed to do that side of it.

Ah, we haven't tagged a release of `civix` since applying the Smarty update there. I get 

> I've updated this to build on the work you did to get to the original goal of the piece of work ie... stop instantiating the Smarty object directly - since as soon as we do that it jumps in on the require_once...  don't know if this will be our final approach but the intention was to separate the loading from the need to register a template to give more options...

Right, OK, we're intervening so that downstream users don't directly instantiate `Smarty`. That can be evaluated on _literal/direct_ basis  or _de-facto/indirect_ basis. Compare:

* By introducing `smarty-v2` mixin, the downstream does not *literally* instantiate `Smarty`. The public contract makes no mention of `CRM_Core_Smarty` or `CRM_Core_Config`.
* If someone enables `smarty-v2@1.0`, then it does in reality (*albeit indirectly*) cause `hook_config` to instantiate `CRM_Core_Smarty` and `Smarty`.

If I'm reading this branch correctly, it still has the same reality - `smarty-v2@1.1` calls `registerTemplateDirectory()` which then instantiates `CRM_Core_Smarty` and `Smarty`.

To reach that original goal, we either need some kind of buffer (or, maybe, some rearranging of the events). I'm sending up a branch to demonstrate what I mean by that...

totten committed:
smarty-v2 - Defer initialization of 'CRM_Core_Smarty'

totten commented:
25553 was passing on its own, so I tacked on a comparable commit here.

totten commented on CRM/Core/Config.php:
```diff
@@ -589,7 +589,10 @@ public function registerTemplateDirectory(array $templateDirectory): void {
     if (($templateDirectory['type'] ?? 'Smarty') === 'Smarty') {
-      CRM_Core_Smarty::singleton()->addTemplateDir($templateDirectory['path']);
+      \Civi::$statics['CRM_Core_Smarty']['PATHS'][] = $templateDirectory['path'];
+      if (CRM_Core_Smarty::$_singleton !== NULL) {
```
@eileenmcnaughton To your concern about when `CRM/Core/Smarty.php` is loaded, perhaps the guard here should be more like

```php
if (class_exists('CRM_Core_Smarty', FALSE) && CRM_Core_Smarty::isStarted()) {
```

The `class_exists(..., FALSE)` means that it will not autoload the class -- it should only tell if you the class has already been loaded.

eileenmcnaughton replied:
@totten hmm - I don't understand how to ensure it's loaded first though - since it was all loading up so quickly
